### PR TITLE
Use javascript libraries from GUIX

### PR DESCRIPTION
**Pull request by BonfaceKilz**
#### Description

This PR replaces all javascript libs with the ones from the guix profile

#### How should this be tested

- First ensure you have the latest [gn2](http://git.genenetwork.org/guix-bioinformatics/guix-bioinformatics.git/) installed in your profile- that will have the recently added js dependencies
- The behaviour in the existing data tables should be the same

#### Any background context you want to provide?

This is part of the effort to decouple javascript from the project

#### What are the relevant pivotal tracker stories?

#396 

#### Screenshots (if appropriate)

N/A

#### Questions

@pjotrp I've changed https://github.com/BonfaceKilz/genenetwork2/blob/20e9676683368877d7eff850179eb1d5812bee03/wqflask/base/trait.py#L433 as a work around for #392. Will that have any unprecedented side effects- I doubt it though- in prod?

### CODE REVIEW

**Comment by pjotrp** [comment]
@zsloan don't merge this yet. I need to update the guix profile with http://git.genenetwork.org/guix-bioinformatics/guix-bioinformatics/commits/branch/master

**Comment by pjotrp** [comment]
Great stuff @BonfaceKilz !

**Comment by pjotrp** [comment]
@BonfaceKilz these packages need to be added to the genenetwork2 package in 
http://git.genenetwork.org/guix-bioinformatics/guix-bioinformatics/src/branch/master/gn/packages/genenetwork.scm#L306. I am now wondering how datatables got included!

**Comment by pjotrp** [comment]
Ah yes, it is in a different branch. http://git.genenetwork.org/guix-bioinformatics/guix-bioinformatics/src/branch/gn-stable/gn/packages/genenetwork.scm#L358

Still need to merge that (you can do that too). How do you test on your own system? Installing packages one at a time? Better test the genenetwork2 'meta' package

**Comment by zsloan** [comment]
I haven't really needed to install anything outside of what I've asked you
to install (like the updated R/qtl).

**Comment by zsloan** [comment]
Regarding the trait.py change, I can't remember if that line was necessary;
there was an issue a while back with the text in the DB being latin1
encoded instead of UTF-8, though it only presents as some search results
with uncommon characters displaying incorrectly. I'll see if I can find
some of the examples from when Rob initially mentioned the problem and use
those to test on my branch.

**Comment by pjotrp** [comment]
> I haven't really needed to install anything outside of what I've asked you to install (like the updated R/qtl).

We are moving the javascript packages into Guix. If you merge this the wqflask static JS links may not work!

**Comment by pjotrp** [comment]
> Regarding the trait.py change, I can't remember if that line was necessary; there was an issue a while back with the text in the DB being latin1 encoded instead of UTF-8, though it only presents as some search results with uncommon characters displaying incorrectly. I'll see if I can find some of the examples from when Rob initially mentioned the problem and use those to test on my branch.

Good idea! With a test we can modify. Note that you spec latin1 in the startup DB connector on production.

**Comment by zsloan** [comment]
Yeah, I just saw and replied to the other github e-mail where Bonface
actually mentioned the error (it wasn't clear to me why the change was made
when I first looked at this e-mail; I didn't know that it was to bypass the
error).

Here's a search for Bonface to test on his branch to see if it's displaying
characters correctly (BXD phenotype trait 20199)  -
http://gn2-zach.genenetwork.org/search?species=mouse&group=BXD&type=Phenotypes&dataset=BXDPublish&search_terms_or=BXD_20199&search_terms_and=&FormID=searchResult

Basically the è in the third author used to be displayed incorrectly
because the code assumed it was getting utf-8 from the DB; after changing
it to latin1 (and having the trait page code assume latin1 encoding when
pulling the data) it was fixed. There might be some other way to deal with
it that I wasn't aware of, though.

**Comment by BonfaceKilz** [comment]
> Ah yes, it is in a different branch. http://git.genenetwork.org/guix-bioinformatics/guix-bioinformatics/src/branch/gn-stable/gn/packages/genenetwork.scm#L358
> 
> Still need to merge that (you can do that too). How do you test on your own system? Installing packages one at a time? Better test the genenetwork2 'meta' package

What I usually do is add a line at the very bottom of the scm file, like say `javascript-datatables` for javascript.scm; and then do: `guix package --install-from-file=javascript.scm --allow-collisions`. I add the `--allow-collisions` option as a work around for conflicts from upstream packages.

**Comment by pjotrp** [comment]
@BonfaceKilz you can just add the dependencies to the genenetwork2 package and install that! Also I use a different profile for such things, e.g.

    env GUIX_PACKAGE_PATH=../guix-bioinformatics guix package -i genenetwork2 -p ~/opt/gn-latest --substitute-urls="http://guix.genenetwork.org https://berlin.guixsd.org https://ci.guix.gnu.org https://mirror.hydra.gnu.org"

Will install genenetwork2 with all its dependencies. That is what we use for deployment.

For the guix.genenetwork.org substitute to work you'll need to add the Guix key to /etc/guix/. It should just work on penguin2, however.

Please add the javascript dependencies to the genenetwork2 package. Next remove the source code from the GN2 source tree and test.

**Comment by pjotrp** [comment]
Oh yes, to activite above profile do

    source ~/opt/gn_latest/etc/profile

**Comment by pjotrp** [comment]
> #### Description
> 
> This PR replaces all dataTable extensions with the ones from the guix profile
> #### How should this be tested
> 
>     * First ensure you have the latest [gn2](http://git.genenetwork.org/guix-bioinformatics/guix-bioinformatics.git/) installed in your profile- that will have the recently added js dependencies

Awesome PR. I'll check today!

>     * The behaviour in the existing data tables should be the same

L&F has changed a bit on GN2 production. Maybe a CSS thing?
 
> #### Any background context you want to provide?
> 
> This is part of the effort to decouple javascript from the project

Woop woop. A long standing wish of mine.

> #### What are the relevant pivotal tracker stories?
> 
> N/A
> #### Screenshots (if appropriate)
> 
> N/A
> #### Questions
> 
> @pjotrp I've changed https://github.com/BonfaceKilz/genenetwork2/blob/20e9676683368877d7eff850179eb1d5812bee03/wqflask/base/trait.py#L433 as a work around for #392. Will that have any unprecedented side effects- I doubt it though- in prod?

@zsloan starts up the mysql client with latin-1. We may need to adapt that to utf-8. I'll check.

**Comment by pjotrp** [comment]
Hi @BonfaceKilz, if you host your GN2 on one of the listed ports it should be visible for us to try. See

       penguin2:$ cat /etc/nginx/sites-enabled/gn2-test.conf

**Comment by BonfaceKilz** [comment]
Thanks. Once I'm able to get over of the gemma-build wall, I'll do that,
and ping you guys.

**Comment by BonfaceKilz** [comment]
@zsloan, @pjotrp and @robwwilliams I'm done replacing the js. @zsloan if you could extract the modified css elsewhere, I could always cherrypick the commits and add them here. I also updated genenetwork2 with the relevant javascript-dependencies,so that should work too. I haven't bundled:
 
  - [-] jquery.cookie.js
  - [-] zxcvbn-async.js
  - [-] bootstrap

@zsloan if you don't mind, could you have a look at it and check if I've missed anything. I've set up the test server here: http://gn2-test2.genenetwork.org/. Wdyt?

**Review by pjotrp — COMMENTED** [review]
Why not use

/gnu/store/p2l1f3s9i2jsklbl34drlfz3kllanlxs-javascript-bootstrap-3.3.0-checkout/docs/examples/non-responsive/non-responsive.css

**Comment by pjotrp** [comment]
Oh it is just an example ;)

**Comment by BonfaceKilz** [comment]
That's because the ones we are using aren't an exact match. It's a
modified version of that ^

**Comment by pjotrp** [comment]
@BonfaceKilz What is the status of this?

**Comment by BonfaceKilz** [comment]
@pjotrp this is done. Let me first sync this with the latest changes from testing and then mark it ready for review.

**Comment by BonfaceKilz** [comment]
@pjotrp, @pjotrp and @robwwilliams Mechanical Rob now checks for a **200 OK** when he fetches js/ css links scraped from some pages. Anything mispackaged should throw a 404(or sth that's not a 200).

<details>
<summary><b><i>Click to see example output</i></b></summary>
<pre>
http://localhost:5004/static/new/javascript/validation.js ==> OK
http://localhost:5004/static/new/javascript/get_covariates_from_collection.js ==> OK
http://localhost:5004/js/DataTables/js/jquery.dataTables.min.js ==> OK
http://localhost:5004/js/jquery-ui/jquery-ui.min.js ==> OK
http://localhost:5004/js/js_alt/parsley.min.js ==> OK
http://localhost:5004/js/nouislider/nouislider.js ==> OK
http://localhost:5004/static/new/javascript/colorbrewer.js ==> OK
http://localhost:5004/js/DataTablesExtensions/plugins/sorting/natural.js ==> OK
http://localhost:5004/js/js_alt/jstat.min.js ==> OK
http://localhost:5004/static/new/javascript/compare_traits_scatterplot.js ==> OK
https://d3js.org/d3.v3.min.js ==> OK
http://localhost:5004/static/new/javascript/search_results.js ==> OK
http://localhost:5004/static/new/javascript/show_trait.js ==> OK
http://localhost:5004/js/DataTablesExtensions/plugins/sorting/scientific.js ==> OK
http://localhost:5004/js/plotly/plotly.min.js ==> OK
http://localhost:5004/static/new/javascript/show_trait_mapping_tools.js ==> OK
http://localhost:5004/js/jquery-cookie/jquery.cookie.js ==> OK
http://localhost:5004/js/DataTables/js/jquery.dataTables.js ==> OK
http://localhost:5004/js/d3-tip/d3-tip.js ==> OK
http://localhost:5004/static/new/javascript/plotly_probability_plot.js ==> OK
http://rf.revolvermaps.com/0/0/8.js?i=526mdlpknyd&m=0&c=ff0000&cr1=ffffff&f=arial&l=33 ==> OK
http://localhost:5004/js/DataTablesExtensions/buttons/js/dataTables.buttons.min.js ==> OK
http://localhost:5004/js/js_alt/underscore.min.js ==> OK
http://localhost:5004/js/jszip/jszip.min.js ==> OK
http://localhost:5004/js/underscore-string/underscore.string.min.js ==> OK
http://localhost:5004/static/new/javascript/stats.js ==> OK
http://localhost:5004/js/js_alt/timeago.min.js ==> OK
http://localhost:5004/js/shapiro-wilk/shapiro-wilk.js ==> OK
http://localhost:5004/twitter/js/twitterFetcher_min.js ==> OK
http://localhost:5004/js/colorbox/jquery.colorbox-min.js ==> OK
http://localhost:5004/static/new/javascript/scatter-matrix.js ==> OK
</pre>
</details> 

Anything under **localhost:5004/js** is packaged in Guix. See the highlighted region for the links that are checked: https://github.com/genenetwork/genenetwork2/pull/394/commits/161a26ecf697608ae3fe44b51f1a1e594e4292e6#diff-0a9b8ec4d0e9f5968041c9a57c17f809R91-R110. In the event that there's a page that should be checked, you could respond here, so that I add them. Other than that, I things are ok imo. I'll put this up on Penguin2 and nudge @pjotrp or @zsloan to have a look(perhaps there are some CSS [issues](https://trello.com/c/LZ4p1K4Y/293-need-to-redo-a-lot-of-css-stuff-appropriately) we haven't noticed yet.

**Comment by BonfaceKilz** [comment]
Also, this PR is big because I deleted alot of js files that was already packaged in GUIX

**Comment by BonfaceKilz** [comment]
@zsloan, @pjotrp and @robwwilliams  I'm going to break down this PR into smaller more modular PRs. That way, we can move everything to guix in a more stable reliable way. I already did most of the heavy lifting, so it would be a matter of harvesting commits from this branch. Each PR will tackle only one JS library. Wdyt?

**Comment by BonfaceKilz** [comment]
Going ahead to close this because of the reason mentioned above :point_up_2:
